src/main/java/com/mongodb/kafka/connect/source/MongoCopyDataManager.java

```diff
@@ -0,0 +1,162 @@
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
+
+import org.apache.kafka.connect.errors.ConnectException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.bson.BsonDocument;
+import org.bson.conversions.Bson;
+
+import com.mongodb.MongoNamespace;
+import com.mongodb.client.MongoClient;
+
+
+/**
+ * Copy Data Manager
+ *
+ * <ol>
+ * <li>Gets all namespaces to copy. eg. A single collection, all collections in a database or all collections in all databases.</li>
+ * <li>For each namespace, submit to the executors an copy existing task.</li>
+ * <li>Each copy existing task, runs an aggregation pipeline on the namespace, to mimic an insert document and adds the results to the
+ *     queue.</li>
+ * <li>The poll method returns documents from the queue.</li>
+ * </ol>
+ */
+class MongoCopyDataManager implements AutoCloseable {
+    private static final Logger LOGGER = LoggerFactory.getLogger(MongoCopyDataManager.class);
+    private final AtomicBoolean closed = new AtomicBoolean();
+    private final AtomicBoolean errored = new AtomicBoolean();
+    private volatile String errorMessage;
+    private final AtomicInteger namespacesToCopy;
+    private final MongoSourceConfig sourceConfig;
+    private final MongoClient mongoClient;
+    private final ExecutorService executor;
+    private final ConcurrentLinkedQueue<BsonDocument> queue = new ConcurrentLinkedQueue<>();
+
+    MongoCopyDataManager(final MongoSourceConfig sourceConfig, final MongoClient mongoClient) {
+        this.sourceConfig = sourceConfig;
+        this.mongoClient = mongoClient;
+
+        String database = sourceConfig.getString(DATABASE_CONFIG);
+        String collection = sourceConfig.getString(COLLECTION_CONFIG);
+
+        List<MongoNamespace> namespaces;
+        if (database.isEmpty()) {
+            namespaces = getCollections(mongoClient);
+        } else if (collection.isEmpty()) {
+            namespaces = getCollections(mongoClient, database);
+        } else {
+            namespaces = singletonList(createNamespace(database, collection));
+        }
+        LOGGER.info("Copying existing data on the following namespaces: {}", namespaces);
+        namespacesToCopy = new AtomicInteger(namespaces.size());
+        executor = Executors.newFixedThreadPool(
+                Math.min(namespacesToCopy.get(), sourceConfig.getInt(COPY_EXISTING_MAX_THREADS_CONFIG)));
+        namespaces.forEach(n -> executor.submit(() -> copyDataFrom(n)));
+    }
+
+    Optional<BsonDocument> poll() {
+        if (errored.get()) {
+            if (!closed.get()) {
+                LOGGER.error(errorMessage);
+                close();
+                throw new ConnectException(errorMessage);
+            }
+            return Optional.empty();
```

Comment:
I think it would be better to just throw an exception if poll is closed after this has been closed.

Reply:
Done